labgrid/driver/rawnetworkinterfacedriver.py

```diff
@@ -35,6 +45,126 @@ def _wrap_command(self, args):
             # keep wrapper and args as-is
             return wrapper + args
 
+    @step(args=["state"])
+    def _set_interface(self, state):
+        """Set interface to given state."""
+        cmd = ["ip", self.iface.ifname, state]
+        cmd = self._wrap_command(cmd)
+        subprocess.check_call(cmd)
+
+    @Driver.check_active
+    def set_interface_up(self):
+        """Set bound interface up."""
+        self._set_interface("up")
+
+    @Driver.check_active
+    def set_interface_down(self):
+        """Set bound interface down."""
+        self._set_interface("down")
+
+    def _get_state(self):
+        """Returns the bound interface's operstate."""
+        if_state = self.iface.extra.get("state")
+        if if_state:
+            return if_state
+
+        cmd = self.iface.command_prefix + ["cat", f"/sys/class/net/{self.iface.ifname}/operstate"]
+        output = processwrapper.check_output(cmd).decode("ascii")
+        if_state = output.strip()
+        return if_state
+
+    @Driver.check_active
+    def get_state(self):
+        """Returns the bound interface's operstate."""
+        return self._get_state()
+
+    @step(title="wait_state", args=["expected_state", "timeout"])
+    def _wait_state(self, expected_state, timeout=60):
+        """Wait until the expected state is reached or the timeout expires."""
+        timeout = Timeout(float(timeout))
+
+        while True:
+            if self._get_state() == expected_state:
+                return
+            if timeout.expired:
+                raise TimeoutError(
+                    f"exported interface {self.iface.ifname} did not go {expected_state} within {timeout.timeout} seconds"
+                )
+            time.sleep(1)
+
+    @Driver.check_active
+    def wait_state(self, expected_state, timeout=60):
+        """Wait until the expected state is reached or the timeout expires."""
+        self._wait_state(expected_state, timeout=timeout)
```

Comment:
Why do you add both a `wait_state()` and the hidden `_wait_state()`? It would be simpler to just merge them.

Reply:
`_wait_state()` is used in `on_activate()`, meaning we can't decorate it with `@Driver.check_active`. That's why I've added both variants.